src/System.Device.Gpio/System/Device/Gpio/Drivers/SunxiDriver.Linux.cs

```diff
@@ -0,0 +1,470 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Device.Gpio.Drivers
+{
+    /// <summary>
+    /// A generic GPIO driver for Allwinner SoCs.
+    /// </summary>
+    /// <remarks>
+    /// This is a generic GPIO driver for Allwinner SoCs.
+    /// It can even drive the internal pins that are not drawn out.
+    /// Before you operate, you must be clear about what you are doing.
+    /// </remarks>
+    public unsafe partial class SunxiDriver : GpioDriver
+    {
+        private IntPtr _gpioPointer0;
+        private IntPtr _gpioPointer1;
+
+        /// <summary>
+        /// CPUX-PORT base address.
+        /// </summary>
+        protected virtual int GpioRegisterOffset0 { get; }
+        /// <summary>
+        /// CPUS-PORT base address.
+        /// </summary>
+        protected virtual int GpioRegisterOffset1 { get; }
+        // final_address = mapped_address + (target_address & map_mask) https://stackoverflow.com/a/37922968
+        private readonly int _mapMask = Environment.SystemPageSize - 1;
+
+        private static readonly object s_initializationLock = new object();
+        private static readonly object s_sysFsInitializationLock = new object();
+        private const string GpioMemoryFilePath = "/dev/mem";
+        private UnixDriver _sysFSDriver;
+        private readonly IDictionary<int, PinMode> _sysFSModes = new Dictionary<int, PinMode>();
+
+        protected SunxiDriver() { }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SunxiDriver"/>.
+        /// </summary>
+        /// <param name="gpioRegisterOffset0">CPUX-PORT base address (This can be find in the corresponding SoC datasheet).</param>
+        /// <param name="gpioRegisterOffset1">CPUS-PORT base address (This can be find in the corresponding SoC datasheet).</param>
+        public SunxiDriver(int gpioRegisterOffset0, int gpioRegisterOffset1)
```

Comment:
nit: might be worth matching the name with params description (i.e.: `cpuxPortBaseAddess` etc)